src/gp_dag.cpp

```diff
@@ -106,6 +109,87 @@ EigenVectorXd GPDAG::BuildUniformQ() const {
   return q;
 }
 
+Node::NodePtrVec GPDAG::GenerateAllTrees() const {
+  std::vector<Node::NodePtrVec> trees_below(NodeCount());
+
+  auto GetSubtrees = [&trees_below](const GPDAGNode *node,
+                                          Node::NodePtrVec &rotated_subtrees,
+                                          Node::NodePtrVec &ordered_subtrees) {
+    for (const bool rotated : {false, true}) {
+      for (const auto &child_id : node->GetLeafward(rotated)) {
+        for (const auto &sub_tree : trees_below.at(child_id)) {
+          rotated ? rotated_subtrees.push_back(sub_tree) :
+          ordered_subtrees.push_back(sub_tree);
+        }
+      }
+    }
+  };
+
+  auto MergeTrees = [](size_t node_id,
+                       Node::NodePtrVec &trees,
+                       Node::NodePtrVec &rotated_subtrees,
+                       Node::NodePtrVec &ordered_subtrees) {
+    for (size_t i = 0; i < rotated_subtrees.size(); i++) {
+      auto &rotated_subtree = rotated_subtrees.at(i);
+      for (size_t j = 0; j < ordered_subtrees.size(); j++) {
+        auto &ordered_subtree = ordered_subtrees.at(j);
+        Node::NodePtr new_tree = Node::Join(ordered_subtree, rotated_subtree, node_id);
+        trees.push_back(new_tree);
+      }
+    }
+  };
+
+  for (const auto &node_id : RootwardPassTraversal()) {
+    const auto &node = GetDagNode(node_id);
+    if (node->IsLeaf()) {
+      trees_below.at(node_id).push_back(Node::Leaf(node_id));
+    } else {
+      Node::NodePtrVec rotated_nodes, ordered_nodes;
+      GetSubtrees(node, rotated_nodes, ordered_nodes);
+      MergeTrees(node_id, trees_below.at(node_id),
+                 rotated_nodes, ordered_nodes);
+    }
+  }
+
+  Node::NodePtrVec trees;
+  IterateOverRootsplitIds([&trees, &trees_below](size_t root_id) {
+    trees.insert(trees.end(), trees_below.at(root_id).begin(), trees_below.at(root_id).end());
+  });
+  
+  Assert(trees.size() == TreeCount(), "The realized number of trees does not match the expected count.");
+
+  return trees;
```

Comment:
Let's call `Polish` on all the trees here.

Reply:
I assign the DAG Ids as the node IDs. And these are used by the caller of this function, in particular, `GetRootedTreeCollection` to look up the GPCSP index. Polish will reassign node IDs so we aren't ready to do that yet here.

Reply:
OK, this makes sense! Let's just signal that to the caller by renaming the method GenerateAllGPNodeIndexedTopologies or something, with a comment clarifying that this is what you are doing.

Reply:
Done.